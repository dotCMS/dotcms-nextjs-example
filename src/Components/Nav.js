import React, { Component } from 'react';
import { Nav as BootstrapNav, NavItem, NavLink, NavbarToggler, Collapse } from 'reactstrap';
import { Link } from 'react-router-dom';

export default class Nav extends Component {
    state = {
        isOpen: false,
        items: []
    };

    async componentDidMount() {
        fetch('/api/v1/nav//?depth=2', {
            headers: {
                DOTAUTH: Buffer.from(
                    `${process.env.REACT_APP_USER_EMAIL}:${process.env.REACT_APP_USER_PASSWORD}`
                ).toString('base64')
            }
        })
            .then(response => response.json())
            .then(data => data.entity.children)
            .then(data => {
                this.setState({
                    ...this.state,
                    items: data
                });
            });
    }

    toggleOpen = e => {
        this.setState({
            ...this.state,
            isOpen: !this.state.isOpen
        });
    };

    render() {
        return (
            <>
                <NavbarToggler onClick={this.toggleOpen} />
                <Collapse isOpen={this.state.isOpen} navbar>
                    <BootstrapNav pills={true}>
                        {this.state.items.map(item => {
                            return (
                                <NavItem key={item.folder}>
                                    <NavLink tag={Link} active={item.href === window.location.pathname} to={item.href}>
                                        {item.title}
                                    </NavLink>
                                </NavItem>
                            );
                        })}
                    </BootstrapNav>
                </Collapse>
            </>
        );
    }
}
