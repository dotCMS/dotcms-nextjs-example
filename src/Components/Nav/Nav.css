div.nav-menu__links {
    flex-grow: unset;
}

button.nav-menu__toggle-button.navbar-toggler {
    color: var(--white);
    border-color: transparent;
}
