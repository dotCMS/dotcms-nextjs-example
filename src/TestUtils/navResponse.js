export const NavResponse = {
    entity: {
        children: [
            {
                href: '/about',
                title: 'About Us',
                folder: '123',
                type: 'htmlpage',
                children: []
            },
            {
                href: '/products',
                title: 'Products',
                folder: '456',
                type: 'htmlpage',
                children: []
            },
            {
                href: '/resources',
                title: 'Resources',
                folder: '789',
                type: 'folder',
                children: [
                    {
                        href: '/videos',
                        title: 'Videos',
                        type: 'htmlpage',
                        folder: null
                    },
                    {
                        href: '/blog',
                        title: 'Blog',
                        type: 'htmlpage',
                        folder: null
                    }
                ]
            }
        ]
    }
};
