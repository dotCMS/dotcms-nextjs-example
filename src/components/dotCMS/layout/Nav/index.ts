export * from './MenuList'
export * from './Nav'
export * from './NavBarSearch'
export * from './NavDropDown'
export * from './NavOption'
export * from './NavSort'
export * from './SocialMediaMenu'
