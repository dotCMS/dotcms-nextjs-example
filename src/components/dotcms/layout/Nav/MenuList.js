import React from 'react';
import RouterLink from '../../../RouterLink';
import { useRouter } from 'next/router';

const getHref = href => {
    const ref = href.split('/').filter(part => part !== 'index').join('/');

    return ref || '/'
}
 
function MenuList({ navigation }) {
    const router = useRouter();

    const routerLinkClassName = (item) => {
        return [
            router.asPath.split('/').filter(Boolean)[0] === item.href.split('/')[1] ? 'active' : '',
            item.children ? 'hasChildren' : ''
        ];
    };

    return (
        <ul className="menu menu__list">
            {navigation.map((item, index) => (
                <li className={routerLinkClassName(item).join(' ')} key={index}>
                    <RouterLink href={getHref(item.href)} key={item.href}>
                        {item.title}
                    </RouterLink>
                    {item.children && (
                        <ul className="submenu">
                            {item.children[0].children.map((child, idx) => {
                                return (
                                    <li key={idx}>
                                        <RouterLink href={getHref(child.href)}>{child.title}</RouterLink>
                                    </li>
                                );
                            })}
                        </ul>
                    )}
                </li>
            ))}
        </ul>
    );
}



export default MenuList;
