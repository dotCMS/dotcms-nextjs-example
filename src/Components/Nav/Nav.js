import React, { Component } from 'react';
import { Nav as BootstrapNav, NavbarToggler, Collapse } from 'reactstrap';
import DotCMSApi from '../../libs/dotcms.api';
import PageContext from '../../PageContext';
import NavOption from '../../Components/Nav/NavOption';
import NavDropDown from './NavDropDown';
import NavSort from './NavSort';
import './Nav.css';

export default class Nav extends Component {
    static contextType = PageContext;
    state = {
        collapsed: false,
        showLogin: false,
        items: []
    };

    componentDidMount() {
        DotCMSApi.request({
            url: `${process.env.REACT_APP_DOTCMS_HOST}/api/v1/nav//?depth=3`
        })
            .then(response => response.json())
            .then(data => data.entity.children)
            .then(data => {
                this.setState({
                    ...this.state,
                    items: data
                });
            })
            .catch(e => console.error(`.catch(${e})`));
    }

    toggleCollapsed = e => {
        this.setState({
            ...this.state,
            collapsed: !this.state.collapsed
        });
    };

    toggleLogin = e => {
        this.setState({
            ...this.state,
            showLogin: !this.state.showLogin
        });
    };

    render() {
        const isEditModeFromDotCMS = this.context && this.context.mode && this.context.mode === 'EDIT_MODE';
        const remoteRendered = this.context && this.context.page && this.context.page.remoteRendered;

        return (
            <>
                {remoteRendered && isEditModeFromDotCMS ? (<NavSort></NavSort>) : ''}
                <NavbarToggler className="nav-menu__toggle-button" onClick={this.toggleCollapsed} >&#9776;</NavbarToggler>
                <Collapse className="nav-menu__links" isOpen={this.state.collapsed} navbar>
                    <BootstrapNav>
                        {this.state.items.map((item) => {
                            if (item.children.length && item.type === 'folder') {
                                return <NavDropDown key={item.hash} options={item} />;
                            } else {
                                return <NavOption key={item.hash} item={item} />;
                            }
                        })}
                    </BootstrapNav>
                </Collapse>
            </>
        );
    }
}
