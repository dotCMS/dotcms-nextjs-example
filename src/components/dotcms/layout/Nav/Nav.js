import React, { useState, useContext } from 'react'
import Image from 'next/image'

import LanguageSelector from '../LanguageSelector'
import MenuList from './MenuList'
import PageContext from '../../../../contexts/PageContext'
import RouterLink from '../../../RouterLink'
import SocialMediaMenu from './SocialMediaMenu'
import useNav from '../../../../hooks/useNav'
import { MainNav, NavMenu } from '../../../../styles/nav/nav.styles'
import { getCurrentLanguage } from '../../../../utilities/dotcms/locale'

export default function Nav() {
  const nav = useNav()
  const [isMenuOpen, setIsMenuOpen] = useState(false)

  const {
    languageProps: { defaultLanguage },
  } = useContext(PageContext)

  const handleOpenMenu = (e) => {
    e.preventDefault()
    setIsMenuOpen(!isMenuOpen)
  }

  return (
    <div className="container">
      <MainNav className="main-nav">
        <RouterLink
          href={
            getCurrentLanguage() && !defaultLanguage
              ? `/${getCurrentLanguage()}`
              : `/`
          }
        >
          <Image
            alt="DotCMS - NextJS example"
            height="41"
            src="/logo.png"
            width="135"
          />
        </RouterLink>
        <NavMenu isOpen={isMenuOpen}>
          <button
            aria-label="button"
            className="hamburger"
            href="#"
            onClick={(e) => handleOpenMenu(e)}
          >
            <Image alt="Open Menu" height="24" src="/menu.svg" width="24" />
          </button>
          <MenuList navigation={nav} />
          <SocialMediaMenu />
          <LanguageSelector />
        </NavMenu>
      </MainNav>
    </div>
  )
}
