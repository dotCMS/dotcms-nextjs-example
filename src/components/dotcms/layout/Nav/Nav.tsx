// Dependencies
import * as React from 'react'
import Image from 'next/image'

// Internals
import { Link } from '@/components'
import { PageContext } from '@/contexts'
import { useNav } from '@/hooks'
import { getCurrentLanguage } from '@/lib/dotCMS'
import { MainNav, NavMenu } from '@/styles/nav/nav.styles'
import { LanguageSelector } from '../LanguageSelector'
import MenuList from './MenuList'
import SocialMediaMenu from './SocialMediaMenu'

export const Nav = () => {
  const nav = useNav()
  const [isMenuOpen, setIsMenuOpen] = React.useState(false)

  const {
    // @ts-ignore TODO: add the correct types
    languageProps: { defaultLanguage },
  } = React.useContext(PageContext)

  const handleOpenMenu = (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault()
    setIsMenuOpen(!isMenuOpen)
  }

  return (
    <div className="container">
      <MainNav className="main-nav">
        <Link
          href={
            getCurrentLanguage() && !defaultLanguage
              ? `/${getCurrentLanguage()}`
              : `/`
          }
        >
          <Image
            alt="DotCMS - NextJS example"
            height="41"
            src="/logo.png"
            width="135"
          />
        </Link>
        <NavMenu isOpen={isMenuOpen}>
          <button
            aria-label="button"
            className="hamburger"
            onClick={(e) => handleOpenMenu(e)}
          >
            <Image alt="Open Menu" height="24" src="/menu.svg" width="24" />
          </button>
          <MenuList navigation={nav} />
          <SocialMediaMenu />
          <LanguageSelector />
        </NavMenu>
      </MainNav>
    </div>
  )
}

export default Nav
