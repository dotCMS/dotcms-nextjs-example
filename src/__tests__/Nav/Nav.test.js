import React from 'react';
import { shallow } from 'enzyme';
import wait from 'waait';
import Nav from '../../Components/Nav/Nav';
import DotCMSApi from '../../libs/dotcms.api';
import toJSON from 'enzyme-to-json';
import { NavResponse } from '../../TestUtils/navResponse';

describe('<Nav />', () => {
    let wrapper;

    beforeEach(async () => {
        process.env.REACT_APP_DOTCMS_HOST = 'http://demo.dotcms.com';
        DotCMSApi.request = jest.fn().mockImplementation(() => {
            return new Promise((resolve, reject) => {
                resolve({
                    status: 200,
                    json: () =>
                        new Promise((resolve, reject) => {
                            resolve(NavResponse);
                        })
                });
            });
        });

        wrapper = await shallow(<Nav />);
    });

    it('should call api correctly', () => {
        expect(DotCMSApi.request).toHaveBeenCalledWith({
            url: 'http://demo.dotcms.com/api/v1/nav//?depth=3'
        });
    });

    it('renders nav correctly', async () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('should expand menu', async () => {
        expect(wrapper.find('Collapse').props().isOpen).toBe(false);

        wrapper.find('NavbarToggler').simulate('click');
        await wrapper.update();
        await wait();

        expect(wrapper.find('Collapse').props().isOpen).toBe(true);
    });
});
