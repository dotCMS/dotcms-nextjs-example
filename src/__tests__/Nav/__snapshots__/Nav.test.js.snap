// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> renders nav correctly 1`] = `
<Fragment>
  <NavbarToggler
    className="nav-menu__toggle-button"
    onClick={[Function]}
    tag="button"
    type="button"
  >
    â˜°
  </NavbarToggler>
  <Collapse
    appear={false}
    className="nav-menu__links"
    enter={true}
    exit={true}
    in={false}
    isOpen={false}
    mountOnEnter={false}
    navbar={true}
    onEnter={[Function]}
    onEntered={[Function]}
    onEntering={[Function]}
    onExit={[Function]}
    onExited={[Function]}
    onExiting={[Function]}
    tag="div"
    timeout={350}
    unmountOnExit={false}
  >
    <Nav
      pills={true}
      tag="ul"
      vertical={false}
    >
      <NavOption
        item={
          Object {
            "children": Array [],
            "folder": "123",
            "hash": 123,
            "href": "/about",
            "title": "About Us",
            "type": "htmlpage",
          }
        }
        key="123"
      />
      <NavOption
        item={
          Object {
            "children": Array [],
            "folder": "456",
            "hash": 1234,
            "href": "/products",
            "title": "Products",
            "type": "htmlpage",
          }
        }
        key="1234"
      />
      <NavDropDown
        key="4343"
        options={
          Object {
            "children": Array [
              Object {
                "folder": null,
                "hash": 4563,
                "href": "/videos",
                "title": "Videos",
                "type": "htmlpage",
              },
              Object {
                "folder": null,
                "hash": 1249,
                "href": "/blog",
                "title": "Blog",
                "type": "htmlpage",
              },
            ],
            "folder": "789",
            "hash": 4343,
            "href": "/resources",
            "title": "Resources",
            "type": "folder",
          }
        }
      />
    </Nav>
  </Collapse>
</Fragment>
`;
